### PR TITLE
Added function .ADDRSIZE to ca65

### PR DESCRIPTION
**Pull request by Movax12**
Hi!
I have a few ideas I want to add to ca65. This one was borne from the need to be able to check the address size of a symbol when creating custom instructions (that have different opcodes - like zeropage or absolute) with macros. This information is known by ca65 if the symbol is constant or imported. This function can give that information to a macro. 

See here for a practical example: http://forums.nesdev.com/viewtopic.php?p=121457#p121457

More to come, but one at a time I think is best.

### CODE REVIEW

**Comment by Movax12** [comment]
I should add: As it is now, this function will not work until the user enables it with the feature command:
.feature addrsize

File: src/ca65/expr.c [hunk 1]
```diff
@@ -629,6 +629,92 @@ static ExprNode* FuncReferenced (void)
 
 
 
+static ExprNode* FuncAddrSize(void)
```
Comment (oliverschmidt) [inline src/ca65/expr.c]: Please adhere to the exsisting coding style.

File: src/ca65/expr.c [hunk 2]
```diff
@@ -629,6 +629,92 @@ static ExprNode* FuncReferenced (void)
...
+
+
+    /* Assume we don't know the size */
+    AddrSize = 0;
+
+    /* Check for a cheap local which needs special handling */
+    if (CurTok.Tok == TOK_LOCAL_IDENT) {
+
+        /* Cheap local symbol */
+        Sym = SymFindLocal(SymLast, &CurTok.SVal, SYM_FIND_EXISTING);
```
Comment (oliverschmidt) [inline src/ca65/expr.c]: Please adhere to the exsisting coding style. The same is true for the other function definitions and call sites too.

File: src/ca65/expr.c [hunk 3]
```diff
@@ -629,6 +629,92 @@ static ExprNode* FuncReferenced (void)
...
+        if (ParentScope == 0) {
+            /* No such scope */
+            SB_Done(&ScopeName);
+            SB_Done(&Name);
+            return GenLiteral0();
+        }
+
+        /* If ScopeName is empty, no explicit scope was specified. We have to
+        * search upper scope levels in this case.
+        */
```
Comment (oliverschmidt) [inline src/ca65/expr.c]: Please adhere to the exsisting coding style, here multiline comment formatting.

File: src/ca65/expr.c [hunk 4]
```diff
@@ -629,6 +629,92 @@ static ExprNode* FuncReferenced (void)
...
+        }
+
+    }
+
+    /* Check if we have a size */
+    /* if we don't know, return it anyway, zero can mean unknown, or uncomment this code for an error
+    if (AddrSize == 0 ) {
+    Error ("Address size of `%m%p%m%p' is unknown", &ScopeName, &Name);
+    }
+    */
```
Comment (oliverschmidt) [inline src/ca65/expr.c]: I have a strong opinion against commented-out code - especially where there's no comment explaining why the code is commented out.

**Comment by oliverschmidt** [comment]
Hi,

First of all thanks for your interest in contributing to cc65@GitHub :-)

I'm in general not exactly keen in adding specific features to the cc65 tools. If however a) the new code doesn't add significant complexity and b) the new feature doesn't (seem to) change the exsisting behaviour when not explicitly using the new feature I'm open to new features.

Anyway I consider it obligatory that the documentation is extended accordingly - which the same pull request. Here that would mean (as far as I understand) a new `.addrsize`section and an addition to the `.feature`section.

Regards,
Oliver

File: src/ca65/expr.c (hunk 4 above)
Comment (greg-king5) [inline src/ca65/expr.c]: Perhaps, it can be a warning (instead of error) message.  Then, that code might not need to be commented out.

Comment (mrdudz) [inline src/ca65/expr.c]: On Monday 20 April 2015, 05:31:11 greg-king5 <notifications@github.com> wrote:
> > +        if (Sym) {
> > +            AddrSize = Sym->AddrSize;
> > +        }
> > +        else {
> > +            Error("Unknown symbol or scope: `%m%p%m%p'",
> > +                &ScopeName, &Name);
> > +        }
> > +
> > +    }
> > +
> > +    /* Check if we have a size */
> > +    /* if we don't know, return it anyway, zero can mean unknown, or
> > uncomment this code for an error +    if (AddrSize == 0 ) {
> > +    Error ("Address size of `%m%p%m%p' is unknown", &ScopeName, &Name);
> > +    }
> > +    */
> 
> Perhaps, it can be a warning (instead of error) message.  Then, that code
> might not need to be commented out.

shouldnt it really be an error? what about taking the address from a non-
constant expression?

-- 

http://www.hitmen-console.org    http://magicdisk.untergrund.net
http://www.pokefinder.org        http://ar.pokefinder.org

Sch�n w�re es, wenn wir wirklich eine Frei-zeitgesellschaft h�tten, dann w�ren 
echtes Bem�hen und Anstrenungen f�r Dinge m�glich, die �ber dem Arbeiten f�rs 
Fressen liegen, insbesondere der Kunst. Mir scheint es eher, als h�tten wir 
eine Arbeits- bzw -losigkeitsgesellschaft; alles, was an Freizeit entsteht 
wird umgemodelt zur Arbeitslosigkeit, und das, was wir an "Freizeitangeboten" 
haben, erscheint mir eher als Symptomkuriererei f�r die, die Arbeit, aber 
keine Zeit haben, und so "das meisste" aus ihrer Zeit machen m�ssen.

Comment (Movax12) [inline src/ca65/expr.c]: After some testing and thinking about it I am sure that if a symbol is valid it should have an internal address size which would be defined in Sym->AddrSize

The only exception is when a symbol is used before it is defined:

   .scope 
          lda foo
          .out .sprintf(" address size of foo: %d", .addrsize(foo))         
    .endscope

In this case it would return 0 and I think it should emit a warning.

**Comment by Movax12** [comment]
I have updated the code to match the coding style and have updated .addrsize to warn if size is not known. Is this a good solution? Documentation updates to follow.

File: doc/ca65.sgml [hunk 5]
```diff
@@ -1278,6 +1278,35 @@ writable.
 Pseudo functions expect their arguments in parenthesis, and they have a result,
 either a string or an expression.
```
Comment (oliverschmidt) [inline doc/ca65.sgml]: One LF missing.

File: doc/ca65.sgml [hunk 6]
```diff
@@ -1278,6 +1278,35 @@ writable.
...
+  	          	        .out .sprinft("Error, address size unknown for symbol %s", .string(foo))
+  	        .endif
+        .endmacro
+  </verb></tscreen>
+
+  This command is new and must be enabled with the <tt/.FEATURE addrsize/ command.
+
+  See: <tt><ref id=".FEATURE" name=".FEATURE"></tt>
+
+
```
Comment (oliverschmidt) [inline doc/ca65.sgml]: One LF too much.

File: doc/ca65.sgml [hunk 7]
```diff
@@ -2648,11 +2683,6 @@ Here's a list of all control commands and a description, what they do:
     overridden. When using this feature, you may also get into trouble if
     later versions of the assembler define new keywords starting with a dot.
 
-  <tag><tt>loose_char_term</tt><label id="loose_char_term"></tag>
-
-    Accept single quotes as well as double quotes as terminators for char
-    constants.
-
```
Comment (oliverschmidt) [inline doc/ca65.sgml]: Why did you remove this section? Looking at feature.c it seems that `loose_char_term`is present.

File: src/ca65/expr.c [hunk 8]
```diff
@@ -629,6 +629,87 @@ static ExprNode* FuncReferenced (void)
...
+            AddrSize = Sym->AddrSize;
+        } else {
+            Error ("Unknown symbol or scope: `%m%p%m%p'",
+                &ScopeName, &Name);
+        }
+
+    }
+
+    if (AddrSize == 0) {
+        Warning(1, "Unknown address size: `%m%p%m%p'",
```
Comment (oliverschmidt) [inline src/ca65/expr.c]: Space missing.

File: src/ca65/expr.c [hunk 9]
```diff
@@ -629,6 +629,87 @@ static ExprNode* FuncReferenced (void)
...
+            Sym = SymFindAny (ParentScope, &Name);
+        } else {
+            Sym = SymFind (ParentScope, &Name, SYM_FIND_EXISTING);
+        }
+        /* If we found the symbol retrieve the size, otherwise complain */
+        if (Sym) {
+            AddrSize = Sym->AddrSize;
+        } else {
+            Error ("Unknown symbol or scope: `%m%p%m%p'",
+                &ScopeName, &Name);
```
Comment (oliverschmidt) [inline src/ca65/expr.c]: Rather no LF here - but if you want then please align `&` with `"`.

File: src/ca65/expr.c [hunk 10]
```diff
@@ -629,6 +629,87 @@ static ExprNode* FuncReferenced (void)
...
+        }
+
+    }
+
+    if (AddrSize == 0) {
+        Warning(1, "Unknown address size: `%m%p%m%p'",
+            &ScopeName, &Name);
```
Comment (oliverschmidt) [inline src/ca65/expr.c]: Rather no LF here - but if you want then please align `&` with `1`.

**Comment by oliverschmidt** [comment]
Thanks for acting on all aspects of the feedback - incl. the demand for doc :-)

Apart from my usual style nitpicking (and my question above) this pull request is ready for merge - if no one objects.

File: doc/ca65.sgml (hunk 7 above)
Comment (Movax12) [inline doc/ca65.sgml]: This was accidental.

**Comment by Movax12** [comment]
I'll fix the requested later today.

File: doc/ca65.sgml [hunk 11]
```diff
@@ -1278,6 +1278,35 @@ writable.
...
+
+  <tscreen><verb>
+        .macro myLDA foo
+
+  	        .if .ADDRSIZE(foo) = 1
+  	          	        ;do custom command based on zeropage addressing:
+  	          	        .byte .A5h, foo
+  	        .elseif .ADDRSIZE(foo) = 2
+  	          	        ;do custom command based on absolute addressing:
+  	          	        .byte .ADh, foo
```
Comment (greg-king5) [inline doc/ca65.sgml]: `.byte` rejects word-sized operands.

File: doc/ca65.sgml [hunk 12]
```diff
@@ -1278,6 +1278,35 @@ writable.
...
+  	        .elseif .ADDRSIZE(foo) = 0
+  	          	        ; no address size define for this symbol:
+  	          	        .out .sprinft("Error, address size unknown for symbol %s", .string(foo))
```
Comment (greg-king5) [inline doc/ca65.sgml]: `.sprinft` is a typo.

File: doc/ca65.sgml [hunk 13]
```diff
@@ -1278,6 +1278,35 @@ writable.
...
+  size of symbol can allow for custom instructions.
+
+  Example:
+
+  <tscreen><verb>
+        .macro myLDA foo
+
+  	        .if .ADDRSIZE(foo) = 1
+  	          	        ;do custom command based on zeropage addressing:
+  	          	        .byte .A5h, foo
```
Comment (greg-king5) [inline doc/ca65.sgml]: Now, it's my turn "to pick a nit". ;-)

That looks, to me, like too much indentation.

File: doc/ca65.sgml (hunk 11 above)
Comment (greg-king5) [inline doc/ca65.sgml]: What are `.A5h` and `.ADh` supposed to be -- ca65 rejects them.  (If you want them to be hex numbers, then the period must be replaced with a zero.)

Comment (Movax12) [inline doc/ca65.sgml]: Thanks for the notes.

File: doc/ca65.sgml (hunk 13 above)
Comment (Movax12) [inline doc/ca65.sgml]: I matched the other entries, though that blank line after the macro declaration should not be there.

**Comment by Movax12** [comment]
Hopefully I have addressed all concerns. The example code in the documentation was tested and works correctly, so all is good (I hope).
